### DISCUSSION TITLE
using setInterval() globally blocks the tests execution

### DISCUSSION THREAD

**Discussion by futin** [avajs/ava#3353] [Q&A]
Hello,

I was trying to migrate from ava@5.1.1 to ava@6.2.0 when I found that only 1 file was processed, and the execution was stuck until the timeout. Below is the config from package.json:

```json 
{
"ava": {
    "concurrency": 1,
    "timeout": "5m",
    "files": [
      "test/unit/**/*.spec.js" (here I have over 10 files)
    ],
    "failFast": false,
    "verbose": true
  }
```

After a couple of hours of debugging, I found out that simple usage of `setInterval` blocks the whole process, and I would like to know what can cause this behavior, as it does not seem expected (especially because the whole execution is blocked, apparently without reason).

### How to reproduce it?

Add this code in any testing directory, as a separate file:

```js
import test from 'ava'
setInterval(() => {}, x) // x can be any value, I used 60 * 60 * 1000, so the code within the `setInterval` was never even executed.

test('simple test', async t => {
  t.pass()
})
```

**Comment by novemberborn** [comment]
`setInterval()` keeps the event loop going. AVA 6 no longer forcibly exits when all tests are done, instead it waits for the event loop to finish. Which it won't because of the interval.

If this code (the `setInterval()` bit) is used in a server-side application it would prevent graceful shutdown, which isn't great.

You can unref the interval so it doesn't stop the event loop from finishing:

```js
setInterval(() => {}, x).unref()
```

Or if it's created in your test, clear it at the end:

```js
const interval = setInterval(x => {}, x)
test.after.always(() => clearInterval(interval))
```

**Comment by futin** [comment]
@novemberborn Thanks for the response. Basically, anything that keeps the event loop going will block the tests, and code refactoring is always good, but we should be given the option to switch between these 2 modes (force exit or not). Is there a way to trigger an old behavior and forcefully exit once the tests are done?

**Reply by novemberborn** [reply]
See https://github.com/avajs/ava/blob/main/docs/08-common-pitfalls.md#timeouts-because-a-file-failed-to-exit
